illumos-utils/src/opte/non_illumos.rs

```diff
@@ -43,3 +70,268 @@ pub fn delete_all_xde_devices(log: &Logger) -> Result<(), Error> {
     slog::warn!(log, "`xde` driver is a fiction on non-illumos systems");
     Ok(())
 }
+
+#[derive(Debug)]
+pub(crate) struct RouteInfo {
+    pub dest: IpCidr,
+    pub target: RouterTarget,
+    pub class: RouterClass,
+}
+
+impl PartialEq for RouteInfo {
+    fn eq(&self, other: &Self) -> bool {
+        if self.dest != other.dest {
+            return false;
+        }
+        match (self.class, other.class) {
+            (RouterClass::System, RouterClass::Custom) => return false,
+            (RouterClass::Custom, RouterClass::System) => return false,
+            (_, _) => {}
+        }
+        match (self.target, other.target) {
+            (RouterTarget::Drop, RouterTarget::Drop) => true,
+            (
+                RouterTarget::InternetGateway(id0),
+                RouterTarget::InternetGateway(id1),
+            ) => id0 == id1,
+            (RouterTarget::Ip(ip0), RouterTarget::Ip(ip1)) => ip0 == ip1,
+            (
+                RouterTarget::VpcSubnet(cidr0),
+                RouterTarget::VpcSubnet(cidr1),
+            ) => cidr0 == cidr1,
+            (_, _) => false,
+        }
+    }
+}
+
+impl RouteInfo {
+    #[allow(dead_code)]
+    pub fn is_system_default_ipv4_route(&self) -> bool {
+        if self.dest
+            != IpCidr::Ip4(Ipv4Cidr::new(
+                Ipv4Addr::ANY_ADDR,
+                0.try_into().unwrap(),
+            ))
+        {
+            return false;
+        }
+        if !matches!(self.target, RouterTarget::InternetGateway(None)) {
+            return false;
+        }
+        matches!(self.class, RouterClass::System)
+    }
+}
+
+impl From<&AddRouterEntryReq> for RouteInfo {
+    fn from(value: &AddRouterEntryReq) -> Self {
+        Self { dest: value.dest, target: value.target, class: value.class }
+    }
+}
+impl From<&DelRouterEntryReq> for RouteInfo {
+    fn from(value: &DelRouterEntryReq) -> Self {
+        Self { dest: value.dest, target: value.target, class: value.class }
+    }
+}
+
+/// Data for one OPTE port
+#[derive(Debug)]
+pub(crate) struct PortData {
+    /// The OPTE-layer information
+    #[cfg_attr(test, allow(dead_code))]
+    pub port: PortInfo,
+    /// The routes for this port. This simulates the router layer.
+    pub routes: Vec<RouteInfo>,
+}
+
+#[derive(Debug)]
+pub(crate) struct State {
+    pub ports: HashMap<String, PortData>,
+    pub underlay_initialized: bool,
+}
+
+const NO_RESPONSE: NoResp = NoResp { unused: 99 };
+static OPTE_STATE: OnceLock<Mutex<State>> = OnceLock::new();
+
+fn opte_state() -> &'static Mutex<State> {
+    OPTE_STATE.get_or_init(|| {
+        Mutex::new(State { ports: HashMap::new(), underlay_initialized: true })
```

Comment:
Should `underlay_initialized` start out `false`?

Reply:
A good question. I thought so, thought that requires the tests to call `Handle::set_xde_underlay()`. I wasn't sure about that, since it seemed like a difference between the test and real implementations.

But saying that out loud, it's not true :) We always have to call this ourselves when the sled-agent starts, so making the tests do it seems more like parity than not.

Thanks, I'll change this, and require the test to set it as well.

Reply:
Done in 2024b245c